source/MaterialXRender/Handlers/ImageHandler.cpp

```diff
@@ -0,0 +1,181 @@
+    if (foundFilePath.isEmpty())
+    {
+        return false;
+    }
+
+    string extension = foundFilePath.getExtension();
+    ImageLoaderMap::reverse_iterator iter;
+    for (iter = _imageLoaders.rbegin(); iter != _imageLoaders.rend(); ++iter)
+    {
+        ImageLoaderPtr loader = iter->second;
+        if (loader && loader->supportedExtensions().count(extension))
+        {
+            bool saved = iter->second->saveImage(foundFilePath, imageDesc, verticalFlip);
+            if (saved)
+            {
+                return true;
+            }
+        }
+    }
+    return false;
+}
+
+bool ImageHandler::acquireImage(const FilePath& filePath, ImageDesc &imageDesc, bool /*generateMipMaps*/, const Color4* /*fallbackColor*/)
+{
+    FilePath foundFilePath = findFile(filePath);
+    if (foundFilePath.isEmpty())
+    {
+        return false;
+    }
+
+    string extension = foundFilePath.getExtension();
+    ImageLoaderMap::reverse_iterator iter;
+    for (iter = _imageLoaders.rbegin(); iter != _imageLoaders.rend(); ++iter)
+    {
+        ImageLoaderPtr loader = iter->second;
+        if (loader && loader->supportedExtensions().count(extension))
+        {
+            bool acquired = loader->acquireImage(foundFilePath, imageDesc, getRestrictions());
+            if (acquired)
+            {
+                return true;
+            }
+        }
+    }
+    return false;
+}
+
+bool ImageHandler::createColorImage(const Color4& color,
+                                    ImageDesc& desc)
+{
+    // Create a solid color image
+    //
+    desc.resourceBuffer = new float[desc.width * desc.height * desc.channelCount];
+    float* pixel = static_cast<float*>(desc.resourceBuffer);
+    for (size_t i = 0; i<desc.width; i++)
+    {
+        for (size_t j = 0; j<desc.height; j++)
+        {
+            for (unsigned int c = 0; c < desc.channelCount; c++)
+            {
+                *pixel++ = color[c];
+            }
+        }
+    }
+    desc.computeMipCount();
+    return true;
+}
+
+bool ImageHandler::bindImage(const string& /*identifier*/, const ImageSamplingProperties& /*samplingProperties*/)
+{
+    return false;
+}
+
+void ImageHandler::cacheImage(const string& identifier, const ImageDesc& desc)
+{
+    if (!_imageCache.count(identifier))
+    {
+        _imageCache[identifier] = desc;
+    }
+}
+
+void ImageHandler::uncacheImage(const string& identifier)
+{
+    _imageCache.erase(identifier);
+}
+
+const ImageDesc* ImageHandler::getCachedImage(const string& identifier)
+{
+    if (_imageCache.count(identifier))
+    {
+        return &(_imageCache[identifier]);
+    }
+    return nullptr;
+}
+
+void ImageHandler::setSearchPath(const FileSearchPath& path)
+{
+    _searchPath = path;
+}
+
+FilePath ImageHandler::findFile(const FilePath& filePath)
+{
+    return _searchPath.find(filePath);
+}
+
+void ImageHandler::deleteImage(ImageDesc& imageDesc)
+{
+    if (imageDesc.resourceBuffer)
+    {
+        free(imageDesc.resourceBuffer);
```

Comment:
Image buffer memory management needs to be improved (as we discussed earlier).

Having `ImageHandler::deleteImage` virtual here doesn't help since it's each individual ImageLoader (not ImageHandler) that knows in what way the buffer was allocated and how it should be deallocated.

A suggestion was to add an `ImageBuffer` base class that owns the buffer memory, and let derived specializations (`OiioImageBuffer`, `StbImageBuffer`, etc.) handle allocation/deallocation of it. Then replace the `void* resourceBuffer` member with an `ImageBufferPtr buffer`, shared pointer instead, which means deallocation is always done and done appropriately for each image buffer type.